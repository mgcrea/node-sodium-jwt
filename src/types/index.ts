export type Header = {
  alg: 'HS512T';
  typ: 'JWT';
};
export type Payload = {
  sub: string | number;
  iat?: number;
  [s: string]: string | number | null | undefined;
};
