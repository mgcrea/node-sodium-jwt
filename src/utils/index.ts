export * from './base64';
export * from './casting';
export * from './error';
