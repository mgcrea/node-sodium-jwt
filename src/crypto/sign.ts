import sodium from 'sodium-native';
import {Header, Payload} from '../types';
import {asBuffer, base64UrlEncode} from '../utils';

export const signSync = (header: Header, payload: Payload, secretKey: Buffer | string): string => {
  const output = Buffer.allocUnsafe(sodium.crypto_auth_BYTES);
  const message = `${base64UrlEncode(header)}.${base64UrlEncode(payload)}`;
  sodium.crypto_auth(output, Buffer.from(message), asBuffer(secretKey, 'base64'));
  return `${message}.${output.toString('base64url')}`;
};
