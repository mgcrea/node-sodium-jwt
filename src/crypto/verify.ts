import sodium from 'sodium-native';
import {Header, Payload} from '../types';
import {asArray, asBuffer, createError} from '../utils';

export const verifySync = (
  message: string,
  secretKeys: (string | Buffer) | (string | Buffer)[]
): {header: Header; payload: Payload; rotated: boolean} => {
  const splitCharIndex = message.lastIndexOf('.');
  if (splitCharIndex === -1) {
    throw createError('MalformedMessageError', 'The message is malformed');
  }
  const signature = Buffer.from(message.slice(splitCharIndex + 1), 'base64url');
  if (signature.length !== sodium.crypto_auth_BYTES) {
    throw createError('SignatureLengthError', 'The signature does not have the required length');
  }
  const cleartext = message.slice(0, splitCharIndex);
  const cleartextBuffer = Buffer.from(cleartext, 'utf8');

  let rotated = false;
  const success = asArray(secretKeys).some((secretKey, index) => {
    const verified = sodium.crypto_auth_verify(signature, cleartextBuffer, asBuffer(secretKey, 'base64'));
    rotated = verified && index > 0;
    return verified;
  });

  if (!success) {
    throw createError('VerifyError', 'Unable to verify');
  }

  const splits = cleartext.split('.');
  if (splits.length !== 2) {
    throw createError('MalformedMessageError', 'The message header/payload is malformed');
  }
  const header = JSON.parse(Buffer.from(splits[0], 'base64url').toString());
  const payload = JSON.parse(Buffer.from(splits[1], 'base64url').toString());
  return {header, payload, rotated};
};
