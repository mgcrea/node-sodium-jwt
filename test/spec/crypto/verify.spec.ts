import {verifySync} from 'src/crypto';

const jwt =
  'eyJhbGciOiJIUzUxMlQiLCJ0eXAiOiJKV1QifQ.eyJzdWIiOjcsImlhdCI6MTY0MTIyNzg2ODYxMn0.a3_bFmeD5vugh3p998QXTdPbAuAJFQzCm-2MfmRVEAg';
const secretKey = '0DfgJOaVrb2quroZavPLp7KJm+hTCBN6hZKnsKQSM+o=';

describe('Verify', () => {
  it('should properly verify a hashed password', async () => {
    const object = verifySync(jwt, [Buffer.from(secretKey, 'base64')]);
    expect(object).toBeDefined();
    expect(Object.keys(object)).toEqual(['header', 'payload', 'rotated']);
    expect(object).toMatchSnapshot();
  });
  it('should properly support a stringified secretKey', async () => {
    const object = verifySync(jwt, secretKey);
    expect(object).toBeDefined();
    expect(Object.keys(object)).toEqual(['header', 'payload', 'rotated']);
    expect(object).toMatchSnapshot();
  });
  it('should properly support a stringified secretKey in an array', async () => {
    const object = verifySync(jwt, [secretKey]);
    expect(object).toBeDefined();
    expect(Object.keys(object)).toEqual(['header', 'payload', 'rotated']);
    expect(object).toMatchSnapshot();
  });
});
