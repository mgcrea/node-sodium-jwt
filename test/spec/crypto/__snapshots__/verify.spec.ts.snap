// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify should properly support a stringified secretKey 1`] = `
Object {
  "header": Object {
    "alg": "HS512T",
    "typ": "JWT",
  },
  "payload": Object {
    "iat": 1641227868612,
    "sub": 7,
  },
  "rotated": false,
}
`;

exports[`Verify should properly support a stringified secretKey in an array 1`] = `
Object {
  "header": Object {
    "alg": "HS512T",
    "typ": "JWT",
  },
  "payload": Object {
    "iat": 1641227868612,
    "sub": 7,
  },
  "rotated": false,
}
`;

exports[`Verify should properly verify a hashed password 1`] = `
Object {
  "header": Object {
    "alg": "HS512T",
    "typ": "JWT",
  },
  "payload": Object {
    "iat": 1641227868612,
    "sub": 7,
  },
  "rotated": false,
}
`;
